File: packages/node/tracep/src/tracer.ts
/**
 * @tracep/sdk – tracer.ts
 *
 * OTel-wrapping Tracer SDK for Node.js.
 *
 * Interface:
 *   const t = new Tracer(endpoint, apiKey, service)
 *   const tid = t.start(name)
 *   t.trace(tid, cls, fn, step, message, { parent?, level? })
 *   await t.end(tid, 'ok' | 'error')
 *   await t.close()   // shutdown provider
 */

import {
  trace,
  context,
  SpanStatusCode,
  type Span,
  type Tracer as OtelTracer,
  type Context,
} from '@opentelemetry/api';
import { NodeTracerProvider, BatchSpanProcessor } from '@opentelemetry/sdk-trace-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { Resource } from '@opentelemetry/resources';
import { SEMRESATTRS_SERVICE_NAME } from '@opentelemetry/semantic-conventions';

// ---------------------------------------------------------------------------
// Internal types
// ---------------------------------------------------------------------------

/** Key used to deduplicate spans within a single trace. */
type SpanKey = string; // `${cls}::${fn}`

interface TraceEntry {
  /** The root span created by start(). */
  root: Span;
  /** OTel context carrying the root span, used as parent for child spans. */
  rootCtx: Context;
  /** Child spans keyed by cls+fn pair. */
  spans: Map<SpanKey, { span: Span; ctx: Context }>;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function spanKey(cls: string, fn: string): SpanKey {
  return `${cls}::${fn}`;
}

// ---------------------------------------------------------------------------
// Tracer
// ---------------------------------------------------------------------------

export class Tracer {
  private readonly provider: NodeTracerProvider;
  private readonly otelTracer: OtelTracer;
  private readonly traces = new Map<string, TraceEntry>();

  /**
   * @param endpoint  Base URL of the OTLP collector, e.g. "http://localhost:4318"
   * @param apiKey    Bearer token forwarded as Authorization header
   * @param service   Value of the service.name OTel resource attribute
   */
  constructor(
    endpoint: string,
    apiKey: string,
    service: string,
  ) {
    const exporter = new OTLPTraceExporter({
      url: `${endpoint.replace(/\/$/, '')}/v1/traces`,
      headers: {
        Authorization: `Bearer ${apiKey}`,
      },
      // Give the exporter up to 10 s before giving up on a single export
      timeoutMillis: 10_000,
    });

    this.provider = new NodeTracerProvider({
      resource: new Resource({
        [SEMRESATTRS_SERVICE_NAME]: service,
      }),
    });

    this.provider.addSpanProcessor(
      new BatchSpanProcessor(exporter, {
        // Buffer up to 50 spans before forcing an export
        maxExportBatchSize: 50,
        // Flush at most once per second
        scheduledDelayMillis: 1_000,
        // Allow up to 512 spans to queue before dropping
        maxQueueSize: 512,
        // Retry budget per export attempt
        exportTimeoutMillis: 10_000,
      })
    );

    this.provider.register();

    this.otelTracer = trace.getTracer(service, '0.1.0');
  }

  // -------------------------------------------------------------------------
  // start()
  // -------------------------------------------------------------------------

  /**
   * Creates a new root span and returns a hex trace-ID string (tid).
   * All subsequent trace()/end() calls use this tid to locate the trace.
   */
  start(name: string): string {
    const root = this.otelTracer.startSpan(name);
    const rootCtx = trace.setSpan(context.active(), root);
    const traceId = root.spanContext().traceId; // 32-char hex

    this.traces.set(traceId, {
      root,
      rootCtx,
      spans: new Map(),
    });

    return traceId;
  }

  // -------------------------------------------------------------------------
  // trace()
  // -------------------------------------------------------------------------

  /**
   * Finds or creates a child span keyed by (cls, fn) and adds a span event.
   *
   * This method is synchronous and non-blocking – spans live in-memory and
   * are flushed asynchronously by the BatchSpanProcessor.
   *
   * @param tid     Trace-ID returned by start()
   * @param cls     Logical class / module name  → code.namespace attribute
   * @param fn      Function / method name       → code.function attribute
   * @param step    Name of the span event
   * @param message Human-readable message added to the event
   * @param opts.parent  tid of a span to use as OTel context parent (optional)
   * @param opts.level   Severity label, e.g. "info" | "warn" | "error"
   */
  trace(
    tid: string,
    cls: string,
    fn: string,
    step: string,
    message: string,
    opts: { parent?: string; level?: string } = {},
  ): void {
    const entry = this.traces.get(tid);
    if (!entry) {
      // Silently drop – never throw from a fire-and-forget tracer
      return;
    }

    const { level = 'info', parent } = opts;
    const key = spanKey(cls, fn);

    // Resolve parent OTel context
    let parentCtx: Context;
    if (parent) {
      // If a parent span-key is provided, look for it in the same trace
      const parentEntry = this.traces.get(parent);
      parentCtx = parentEntry ? parentEntry.rootCtx : entry.rootCtx;
    } else {
      parentCtx = entry.rootCtx;
    }

    // Reuse an existing span for this cls+fn pair, or create a new one
    if (!entry.spans.has(key)) {
      const span = context.with(parentCtx, () =>
        this.otelTracer.startSpan(`${cls}.${fn}`, {
          attributes: {
            'code.namespace': cls,
            'code.function': fn,
          },
        }),
      );
      const spanCtx = trace.setSpan(parentCtx, span);
      entry.spans.set(key, { span, ctx: spanCtx });
    }

    const { span } = entry.spans.get(key)!;

    // Add the event – this is an in-memory operation, never blocks
    span.addEvent(step, {
      message,
      level,
    });
  }

  // -------------------------------------------------------------------------
  // end()
  // -------------------------------------------------------------------------

  /**
   * Sets status on all spans, ends them in reverse creation order, then
   * force-flushes the provider so the batch exporter ships the data.
   *
   * @param tid    Trace-ID returned by start()
   * @param status 'ok' | 'error'
   */
  async end(tid: string, status: 'ok' | 'error'): Promise<void> {
    const entry = this.traces.get(tid);
    if (!entry) return;

    const otelStatus =
      status === 'ok'
        ? { code: SpanStatusCode.OK }
        : { code: SpanStatusCode.ERROR, message: 'trace ended with error' };

    // End child spans first (LIFO)
    const children = [...entry.spans.values()].reverse();
    for (const { span } of children) {
      span.setStatus(otelStatus);
      span.end();
    }

    // End root span
    entry.root.setStatus(otelStatus);
    entry.root.end();

    this.traces.delete(tid);

    // Force-flush is non-blocking from the caller's perspective – the await
    // here only waits for the in-process export to finish, not for a network
    // round-trip to complete (BatchSpanProcessor handles retries internally).
    try {
      await this.provider.forceFlush();
    } catch {
      // Silently drop – never surface exporter errors to callers
    }
  }

  // -------------------------------------------------------------------------
  // close()
  // -------------------------------------------------------------------------

  /**
   * Gracefully shuts down the TracerProvider, flushing all pending spans.
   * Call this once at process exit.
   */
  async close(): Promise<void> {
    // End any still-open traces as errors
    for (const [tid] of this.traces) {
      await this.end(tid, 'error');
    }
    try {
      await this.provider.shutdown();
    } catch {
      // Silently ignore shutdown errors
    }
  }
}

File: packages/rust/tracep/src/tracer.rs
//! Core [`Tracer`] implementation.
//!
//! Every public method is fire-and-forget: the caller is never blocked by
//! network I/O.  All OTel flushing happens via a Tokio runtime that is owned
//! by the [`Tracer`] instance.

use std::{
    collections::HashMap,
    error::Error,
    sync::{Arc, Mutex},
    time::Duration,
};

use opentelemetry::{
    global,
    trace::{
        Span, SpanKind, Status, TraceContextExt, Tracer as OtelTracer,
    },
    Context, KeyValue,
};
use opentelemetry_otlp::WithExportConfig;
use opentelemetry_sdk::{
    runtime,
    trace::{BatchConfig, Config},
};
use tokio::runtime::Runtime;

// ── internal span book-keeping ────────────────────────────────────────────────

/// One active trace: the root span plus every child span created by `trace()`.
struct TraceEntry {
    /// Root span (created by `start()`).
    root_span: opentelemetry_sdk::trace::Span,
    /// Child spans keyed by `"class::function"`.
    child_spans: HashMap<String, opentelemetry_sdk::trace::Span>,
}

// ── public API ────────────────────────────────────────────────────────────────

/// Fire-and-forget OTel tracer.
///
/// Construct once with [`Tracer::new`], then call [`Tracer::start`] /
/// [`Tracer::trace`] / [`Tracer::end`] from any thread.  All spans are
/// exported via OTLP/HTTP JSON to `<endpoint>/v1/traces`.
pub struct Tracer {
    /// OTel tracer.
    otel_tracer: opentelemetry_sdk::trace::Tracer,
    /// All in-flight traces, keyed by trace-id hex string.
    traces: Arc<Mutex<HashMap<String, TraceEntry>>>,
    /// Dedicated Tokio runtime used for async force-flush calls.
    rt: Runtime,
}

impl Tracer {
    // ── constructor ──────────────────────────────────────────────────────────

    /// Create a new [`Tracer`].
    ///
    /// # Arguments
    ///
    /// * `endpoint` – Base URL of the OTLP collector, e.g. `"http://localhost:4318"`.
    ///   `/v1/traces` is appended automatically.
    /// * `api_key`  – Bearer token sent as `Authorization: Bearer <api_key>`.
    /// * `service`  – Logical service name embedded in every span.
    pub fn new(endpoint: &str, api_key: &str, service: &str) -> Result<Self, Box<dyn Error>> {
        // Build the Tokio runtime first; we need it for the OTLP exporter's
        // reqwest HTTP client.
        let rt = Runtime::new()?;

        let traces_url = format!("{}/v1/traces", endpoint.trim_end_matches('/'));
        let auth_header = format!("Bearer {}", api_key);

        // Build the OTLP HTTP exporter.
        let exporter = opentelemetry_otlp::new_exporter()
            .http()
            .with_endpoint(&traces_url)
            .with_headers({
                let mut h = HashMap::new();
                h.insert("Authorization".to_string(), auth_header);
                h
            })
            .with_timeout(Duration::from_secs(10));

        // Batch processor configuration: buffer up to 50 spans, flush every 1 s.
        let batch_config = BatchConfig::default()
            .with_max_export_batch_size(50)
            .with_scheduled_delay(Duration::from_secs(1));

        // Build the TracerProvider and install it inside the Tokio runtime.
        let otel_tracer = rt.block_on(async {
            opentelemetry_otlp::new_pipeline()
                .tracing()
                .with_exporter(exporter)
                .with_trace_config(
                    Config::default().with_resource(
                        opentelemetry_sdk::Resource::new(vec![KeyValue::new(
                            "service.name",
                            service.to_string(),
                        )]),
                    ),
                )
                .with_batch_config(batch_config)
                .install_batch(runtime::Tokio)
        })?;

        // Retrieve the provider (which is an Option).
        let provider = otel_tracer.provider().ok_or("OTel tracer has no provider")?;

        // Register as the global provider so `global::tracer()` works if
        // callers mix APIs; not strictly required for our SDK but good practice.
        global::set_tracer_provider(provider);

        Ok(Self {
            otel_tracer,
            traces: Arc::new(Mutex::new(HashMap::new())),
            rt,
        })
    }

    // ── start ────────────────────────────────────────────────────────────────

    /// Start a new root trace.
    ///
    /// Returns the trace-id as a lower-case hex string (32 characters).
    pub fn start(&self, name: &str) -> String {
        let span = self
            .otel_tracer
            .span_builder(name.to_string())
            .with_kind(SpanKind::Internal)
            .start(&self.otel_tracer);

        let trace_id = span.span_context().trace_id();
        let tid = format!("{:032x}", u128::from_be_bytes(trace_id.to_bytes()));

        let entry = TraceEntry {
            root_span: span,
            child_spans: HashMap::new(),
        };

        self.traces
            .lock()
            .expect("traces mutex poisoned")
            .insert(tid.clone(), entry);

        tid
    }

    // ── trace ────────────────────────────────────────────────────────────────

    /// Add a span event to the trace identified by `tid`.
    ///
    /// * `cls`      – `code.namespace` attribute value (class / module name).
    /// * `function` – `code.function` attribute value.
    /// * `step`     – Name of the OTel span event.
    /// * `message`  – `message` attribute on the span event.
    /// * `parent`   – Optional parent span key (`"class::function"`).  When
    ///                supplied the newly created span is a child of that span.
    /// * `level`    – Log level string, defaults to `"info"`.
    ///
    /// If `tid` is not recognised the call is silently ignored.
    pub fn trace(
        &self,
        tid: &str,
        cls: &str,
        function: &str,
        step: &str,
        message: &str,
        parent: Option<&str>,
        level: Option<&str>,
    ) {
        let level = level.unwrap_or("info").to_string();
        let span_key = format!("{}::{}", cls, function);
        let step = step.to_string();
        let message = message.to_string();
        let cls = cls.to_string();
        let function = function.to_string();

        let mut guard = self.traces.lock().expect("traces mutex poisoned");

        let entry = match guard.get_mut(tid) {
            Some(e) => e,
            None => return, // unknown tid → silently ignore
        };

        // Resolve the OTel parent context.
        let parent_cx: Context = if let Some(parent_key) = parent {
            if let Some(parent_span) = entry.child_spans.get(parent_key) {
                // Attach the parent span's context so the child is linked.
                Context::current().with_remote_span_context(parent_span.span_context().clone())
            } else {
                // Fall back to root span context.
                Context::current().with_remote_span_context(entry.root_span.span_context().clone())
            }
        } else {
            Context::current().with_remote_span_context(entry.root_span.span_context().clone())
        };

        // Find or create the child span for this class+function pair.
        if !entry.child_spans.contains_key(&span_key) {
            let child_span = self
                .otel_tracer
                .span_builder(span_key.clone())
                .with_kind(SpanKind::Internal)
                .start_with_context(&self.otel_tracer, &parent_cx);

            entry.child_spans.insert(span_key.clone(), child_span);
        }

        // Set attributes and add the event on the child span.
        if let Some(span) = entry.child_spans.get_mut(&span_key) {
            span.set_attribute(KeyValue::new("code.namespace", cls));
            span.set_attribute(KeyValue::new("code.function", function));
            span.add_event(
                step,
                vec![
                    KeyValue::new("message", message),
                    KeyValue::new("level", level),
                ],
            );
        }
    }

    // ── end ──────────────────────────────────────────────────────────────────

    /// Finish the trace identified by `tid`.
    ///
    /// * `status` – `"ok"` sets [`Status::Ok`]; anything else sets
    ///              [`Status::Error`].
    ///
    /// All child spans are ended first, then the root span.  A force-flush
    /// is requested so buffered spans reach the exporter promptly.
    ///
    /// If `tid` is not recognised the call is silently ignored.
    pub fn end(&self, tid: &str, status: &str) {
        let entry = {
            let mut guard = self.traces.lock().expect("traces mutex poisoned");
            guard.remove(tid)
        };

        let mut entry = match entry {
            Some(e) => e,
            None => return,
        };

        let otel_status = if status.eq_ignore_ascii_case("ok") {
            Status::Ok
        } else {
            Status::error(status.to_string())
        };

        // End child spans.
        for (_, mut span) in entry.child_spans.drain() {
            span.set_status(otel_status.clone());
            span.end();
        }

        // End root span.
        entry.root_span.set_status(otel_status);
        entry.root_span.end();

        // Force-flush so the batch processor ships queued spans immediately.
        let provider_opt = self.otel_tracer.provider();
        if let Some(provider) = provider_opt {
            self.rt.spawn(async move {
                // Retry loop: 3 attempts with exponential back-off.
                let mut delay = Duration::from_millis(100);
                for attempt in 0..3u32 {
                    match provider.force_flush() {
                        results if results.iter().all(|r| r.is_ok()) => break,
                        _ => {
                            if attempt < 2 {
                                tokio::time::sleep(delay).await;
                                delay *= 2;
                            }
                            // 3rd failure → drop silently
                        }
                    }
                }
            });
        }
    }

    // ── close ────────────────────────────────────────────────────────────────

    /// Gracefully shut down the tracer.
    ///
    /// Flushes remaining spans and shuts down the OTel provider.  Call once
    /// at application exit.
    pub fn close(&self) {
        if let Some(provider) = self.otel_tracer.provider() {
            // Block until flush completes (within 5 s timeout).
            self.rt.block_on(async move {
                let _ = tokio::time::timeout(Duration::from_secs(5), async {
                    let _ = provider.force_flush();
                })
                .await;
            });
        }
        global::shutdown_tracer_provider();
    }
}

// Safety: `Tracer` can be shared across threads. The `Arc<Mutex<…>>` guards
// mutable state; the OTel provider and tracer are internally `Send + Sync`.
unsafe impl Send for Tracer {}
unsafe impl Sync for Tracer {}
